packages/session/src/index.ts

```diff
@@ -119,14 +125,17 @@ export function createSessionMiddleware(options: SessionOptions) {
 
       sessionPersisted = true;
 
-      const isExistingSession = cookieSessionId && cookieSessionId === req.session.id;
+      // If this is a new session, we would have already persisted it to the
+      // store, so we don't need to take that into consideration here.
+      //
+      // If the hash of the session data changed, we'll unconditionally persist
+      // the updated data to the store. However, if the hash didn't change, we
+      // only want to persist it if the expiration changed *and* if we can set
+      // a cookie to reflect the updated expiration date.
       const hashChanged = hashSession(req.session) !== originalHash;
       const didExpirationChange =
         originalExpirationDate.getTime() !== req.session.getExpirationDate().getTime();
-      if (
-        (hashChanged && isExistingSession) ||
-        (canSetCookie && (!isExistingSession || didExpirationChange))
-      ) {
+      if (hashChanged || (didExpirationChange && canSetCookie)) {
```

Comment:
Aside: I'm reasonably confident that I can remove all the logic around `canSetCookie`. We should be able to set cookies unconditionally, even on subdomains, since we can set the `domain` attribute for cookies. I'll leave that for another PR though.

Reply:
👀